Bram-Hub/LEGUP
# Bugfix 549

## Description

Allows direct rule transitions that have no changes to be considered valid.
Prevents modification of board elements that would affect existing future case rules.

Specifics:
- model/rules/DirectRule.java
  - Adds a case allowing transitions with no modified data
- model/tree/TreeTransition.java.propogateChange()
  - Propagation now updates modified data arrays of child transitions
  - Other minor fixes
- model/rules/CaseRule.java
  - Added function dependentElements(), which specifies elements to be made unmodifiable
    - Added default implementation and overrides for case rules I thought needed them
- model/gameboard/PuzzleElement.java
  - Added .caseDepended attribute, which counts the number of case rules depending on this element
- model/tree/TreeTransition.java
  - Added .selection attribute, which, for case rule transitions, tracks the selected puzzle element
    - This is neccessary to call dependentElements() given only a transition
    - Side Note: this should help in streamlining checkRule() implementations for case rules
      - Replaced skyscrapers-specific implementation of this idea with the new attribute
        -  CellForNumberCaseRule.java, SkyscrapersClue.java, SkyscrapersBoard.java
 - history/AutoCaseRuleCommand.java
   - Set selection attribute of the created transitions
 - proofeditorui/treeview/TreeView.java
   - Added code to update casesDepended for all puzzleElements in the tree whenever:
     - The first transition of a case rule is added
     - The last transition of a case rule is removed
     - A node containing a case rule is added
     - A node containing a case rule is removed
   - If the Case has been added
     -  All dependentElements in all ancestors are incremented and set to unmodifiable
   - If the Case has been removed
     - All dependentElements in all ancestors are decremented
     - If dependentElements is now 0, sets modifiability to what it should be:
       - Modifiable if started modifiable and not modified by any ancestors

Closes #(549)

## Type of change

- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improvement to an already existing feature)

## How Has This Been Tested?

Test using Lightup.SatisfyNumber Case Rule:

1. Open 7x7 Easy 69495
2. Apply Finish With Empty to the top-left 0 
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/e06d89a5-5fbd-4a55-b7ec-122abd0f58e1" width="50%" height="50%">    

3. Edit this transition, and notice that when removing all changes, the transition is still valid 
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/635d5c82-10de-4eeb-8ebe-905ea46ba91d" width="50%" height="50%">

4. Apply case rule to the left-most 2. Note that transition 2 cannot be modified anywhere placing a light would light up a square adjacent to the left-most 2, invalidating a case. (empty placed where possible in this screenshot)
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/52074780-f87c-401e-add7-db79e5eb5bd6" width="50%" height="50%">    

5. From the node shown after transition 4, call Finish With Empty again, filling in the spots now unmodifiable by transition 2. Note that all cells not previously modified are modifiable.
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/61cd2fe2-4249-4ab8-b9ff-796e8be701f9" width="50%" height="50%">    

6. From node 5, Apply case rule to the 2 diagonal to the previous. Note that transition 5 is unmodifiable where placing a light would affect the new 2, and transition 2 cannot be modified where it would affect either 2. (light placed where possible in these screenshots)
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/2b0725d8-48fe-4dfd-8e38-fe0e27cb961b" width="50%" height="50%">  
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/05efea51-0da0-4e55-bbcf-ac9f66254a4d" width="50%" height="50%">     

7. Delete each transition after node 5 individually. Note that modifiability remains the same until the final transition is deleted, at which point we return to the state after step 5.
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/50953bf2-2a2c-4b70-b8e0-b38be7ca828e" width="50%" height="50%">    

8. Undo these deletions. Note that after the first undo, modifiability returns to the state after step 6.
9. Delete node 2. Note modifiability of transition 2 is the same after step 2.
<img src="https://github.com/Bram-Hub/LEGUP/assets/98851950/c16825dc-06ab-4b35-94bb-430abc56b92f" width="50%" height="50%">    

10. Undo this deletion. Note modifiability returns to the state after step 6.


## Checklist:

- [x ] I have performed a self-review of my code
